armonik-client/src/main/java/fr/aneo/armonik/client/internal/grpc/observers/BlobCompletionEventObserver.java

```diff
@@ -0,0 +1,182 @@
+                                     Map<BlobId, BlobHandle> pending,
+                                     CompletableFuture<Void> completion,
+                                     BlobCompletionListener listener) {
+    this.sessionId = sessionId;
+    this.pending = pending.entrySet().stream().collect(Collectors.toConcurrentMap(entry -> entry.getKey().asString(), Map.Entry::getValue));
+    this.remaining = new AtomicInteger(pending.size());
+    this.completion = completion;
+    this.listener = listener;
+  }
+
+  @Override
+  public void onNext(EventSubscriptionResponse resp) {
+    switch (resp.getUpdateCase()) {
+      case RESULT_STATUS_UPDATE ->
+        handleEvent(resp.getResultStatusUpdate().getResultId(), resp.getResultStatusUpdate().getStatus());
+      case NEW_RESULT ->
+        handleEvent(resp.getNewResult().getResultId(), resp.getNewResult().getStatus());
+      case UPDATE_NOT_SET ->
+      { /* ignore */ }
+    }
+
+    if (remaining.get() == 0 && !completion.isDone()) {
+      completion.complete(null);
+    }
+  }
+
+  @Override
+  public void onError(Throwable throwable) {
+    log.atError().addKeyValue("sessionId", sessionId.asString())
+       .addKeyValue("error", throwable.getClass().getSimpleName())
+       .setCause(throwable)
+       .log("Blob Event Stream failed");
+  }
+
+  @Override
+  public void onCompleted() {
+    if (remaining.get() == 0 && !completion.isDone()) {
+      completion.complete(null);
+    }
+  }
+
+  private void handleEvent(String blobId, ResultStatus status) {
+    switch (status) {
+      case RESULT_STATUS_COMPLETED -> {
+        final var blobHandle = pending.remove(blobId);
+        if (blobHandle != null) {
+          handleCompleted(blobHandle);
+        }
+      }
+      case RESULT_STATUS_ABORTED -> {
+        final var blobHandle = pending.remove(blobId);
+        if (blobHandle != null) {
+          handleAborted(blobHandle);
+        }
+      }
+      default -> { /* ignore */ }
+    }
+  }
+
+  private void handleAborted(BlobHandle blobHandle) {
+    blobHandle.deferredBlobInfo()
+              .thenAccept(blobInfo -> {
+                try {
+                  listener.onError(new BlobError(blobInfo, null)); //TODO define exception
+                } catch (Throwable throwable) {
+                  log.atError().addKeyValue("sessionId", sessionId.asString())
+                     .addKeyValue("blobInfo", blobInfo)
+                     .addKeyValue("error", throwable.getClass().getSimpleName())
+                     .setCause(throwable)
+                     .log("BlobCompletionListener throws an Exception");
+                } finally {
+                  if (remaining.decrementAndGet() == 0 && !completion.isDone()) {
+                    completion.complete(null);
+                  }
+                }
+              });
+  }
+
+  private void handleCompleted(BlobHandle blobHandle) {
+    blobHandle.deferredBlobInfo().thenCompose(blobInfo ->
+      blobHandle.downloadData()
```

Comment:
Is there a retry mechanism ?

Reply:
Specifically, `blobHandle.downloadData()` calls `ResultsStub.downloadResultData()`, which relies on gRPC’s automatic transparent retries for transient transport errors (like a dropped connection before headers are received). Beyond that, no retry policy is currently configured, so application-level retries would need to be added explicitly later.

Reply:
Ok we will need to implement the retry policy for streams later on